src/api_types.jl

```diff
@@ -218,8 +238,8 @@ const H5P_LINK_ACCESS      = _read_const(:H5P_CLS_LINK_ACCESS_ID_g)
 # Reference constants
 const H5R_OBJECT         = 0
 const H5R_DATASET_REGION = 1
-const H5R_OBJ_REF_BUF_SIZE      = 8
-const H5R_DSET_REG_REF_BUF_SIZE = 12
+const H5R_OBJ_REF_BUF_SIZE      = 8  # == sizeof(hobj_ref_t)
```

Comment:
Am I looking at the right code
https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5Rpublic.h#29
?

Reply:
You want the v1.10 branch:

https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5Rpublic.h?at=refs%2Fheads%2Fhdf5_1_10#29-39

and

https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5Rpublic.h?at=refs%2Fheads%2Fhdf5_1_10#53-64

The comments are not a literal translation of how the C headers define the constants; they were meant to just clarify that those correspond to the type sizes.

Reply:
I'm a little confused on some of the changes here. 

Shouldn't `const hobj_ref_t = haddr_t`, i.e. `UInt64` but it seems like here it's instead being represented by  an 'equivalent' representation consistent of `NTuple{8,UInt8}` 

Also I'm having a hard time understanding what this means?
```
    # UInt64 and not UInt8 to get 8-byte alignment
```
Yet the comment is using the ntuple.

Reply:
The first one was just for symmetry with the other types — they're supposed to be opaque data structures, so the actual representation shouldn't matter. I could just change that back to `UInt64` if you prefer, though.

The second question just means that the element type of the tuple is `UInt64` rather than `UInt8` — [this gist](https://gist.github.com/jmert/e60e6f0b49eed89b527a1cabddc4d23b) I linked to in an earlier comment suggests LLVM doesn't maintain the full 8-byte alignment if the tuple is made of `UInt8`s while it does if the elements are `UInt64`s, so it's just to try to mirror the reason the C code uses a `union`.

Reply:
I guess `UInt64` does impose an alignment, so that could be a semantic change caused by changing to a tuple of `UInt8`s.

Reply:
Running the code in your gist I obtain:
![image](https://user-images.githubusercontent.com/4319522/95121245-c5909e00-071c-11eb-9d7b-ab5faef229d6.png)

```
julia> versioninfo()
Julia Version 1.6.0-DEV.1117
Commit 36effbe10a (2020-10-02 17:38 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-10.0.1 (ORCJIT, icelake-client)
Environment:
  JULIA_EDITOR = code.cmd
  JULIA_NUM_THREADS = 8
```

Reply:
Oh I see, how do we know what's the correct alignment to use?

Reply:
I think you just have to follow C semantics. Since that type is typedef'd to `haddr_t`, it'll have that alignment if e.g. a user were to embed it within a struct, so we should make sure to duplicate that property.

Reply:
I see, yeah these alignment issues are interesting and I think good to resolve, but I don't think they will impact any user-facing code or internal usage, just yet.